### PR TITLE
Add key property to Role entity

### PR DESCRIPTION
**Pull request by nnatter**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | yes
| Deprecations? | no
| Related issues/PRs | #3675
| License | MIT

#### What's in this PR?

This PR adds a `key` property to the `Role` entity. The `key` property is implemented similar to the `key` property of the `Category` entity. 

Furthermore, it adjust the logic to use the `key` property instead of the `name` property for generating the identifier of a role. The identifier of a role is used for restricting access of a user in the [Symfony security configuration](https://symfony.com/doc/current/security.html#securing-url-patterns-access-control).

#### Why?

At the moment, we use the `name` property for generating the identifier of a role. This means that changing the `name` of a role might break the Symfony security configuration of the project. See #3675 for more information.

I am aware that this is not a perfect solution, but it is a solution that improves the current situation and is implementable with relatively low effort and small BC breaks. We can still implement something like `System Roles` in the future.

#### BC Breaks/Deprecations

See `UPGRADE.MD`

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -2,6 +2,25 @@
...
+keep backwards compatibility:
+
+```sql
+DROP INDEX UNIQ_13B749A05E237E06 ON se_roles;
+ALTER TABLE se_roles ADD `role_key` VARCHAR(60) DEFAULT NULL;
+UPDATE se_roles SET `role_key` = `name` WHERE `role_key` IS NULL;
+CREATE UNIQUE INDEX UNIQ_13B749A03EF22FDB ON se_roles (role_key);
+```
+
+### Remove getRole() method of Role entity
```
Comment (alexander-schranz) [inline UPGRADE.md:19]: We should just deprecate the getRole. We should not make life heavier as it should be.

File: src/Sulu/Bundle/SecurityBundle/Entity/Role.php [hunk 2]
```diff
@@ -93,14 +98,9 @@ public function getId()
...
 
-    public function getRole()
-    {
-        return 'ROLE_SULU_' . strtoupper($this->name);
-    }
-
     public function getIdentifier()
     {
-        return 'ROLE_SULU_' . strtoupper($this->getName());
+        return 'ROLE_SULU_' . strtoupper($this->getKey());
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Entity/Role.php:103]: I would make the key nullable and fallback to the name makes the migration simplier and doesn't crash projects where developer did just `doctrine:schema:update` or where developer did build custom logic with ROLE_USER_{NAME}.

```suggestion
$key = $this->getKey();

if (!$key) {
    $key = $this->getName();
}

return 'ROLE_SULU_' . strtoupper($key);
```

File: src/Sulu/Bundle/SecurityBundle/Resources/config/doctrine/Role.orm.xml [hunk 3]
```diff
@@ -8,7 +8,8 @@
             <generator strategy="AUTO"/>
         </id>
 
-        <field name="name" type="string" column="name" length="60" unique="true"/>
+        <field name="name" type="string" column="name" length="60"/>
+        <field name="key" type="string" length="60" column="role_key" nullable="true" unique="true"/>
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Resources/config/doctrine/Role.orm.xml:12]: would make the key nullable

File: src/Sulu/Bundle/SecurityBundle/Resources/config/forms/role_details.xml [hunk 4]
```diff
@@ -15,7 +15,13 @@
...
         </property>
 
-        <property name="system" type="single_select" mandatory="true" colspan="3" spaceAfter="9">
+        <property name="key" type="text_line" colspan="9">
+            <meta>
+                <title>sulu_admin.key</title>
+            </meta>
+        </property>
+
+        <property name="system" type="single_select" mandatory="true" colspan="3">
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Resources/config/forms/role_details.xml:24]: would move system before key because key is optional

File: src/Sulu/Bundle/SecurityBundle/Entity/Role.php [hunk 5]
```diff
@@ -127,6 +127,18 @@ public function getName()
...
     }
 
+    public function getKey()
+    {
+        return $this->key;
+    }
+
+    public function setKey($key)
+    {
+        $this->key = $key;
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Entity/Role.php:137]: Can roles with spaces be configured in the firewall e.g. `ROLE_TEST USER`?

File: UPGRADE.md [hunk 6]
```diff
@@ -2,6 +2,25 @@
...
+### Add key property to Role entity
+
+To allow for referencing `Role` entities by a human readable string, a `key` property was added to the `Role` entity.
+Furthermore, the `key` property is used for generating the identifier of a role instead of the `name` property now.
+The following statements update the database to include the new property and set the key of each existing Role to 
+keep backwards compatibility:
+
+```sql
+DROP INDEX UNIQ_13B749A05E237E06 ON se_roles;
+ALTER TABLE se_roles ADD `role_key` VARCHAR(60) DEFAULT NULL;
```
Comment (alexander-schranz) [inline UPGRADE.md:14]: would make the key nullable

File: UPGRADE.md [hunk 7]
```diff
@@ -2,6 +2,25 @@
...
+UPDATE se_roles SET `role_key` = `name` WHERE `role_key` IS NULL;
```
Comment (alexander-schranz) [inline UPGRADE.md:15]: this would not be needed when making it nullable

File: src/Sulu/Bundle/SecurityBundle/Entity/Role.php (hunk 2 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/SecurityBundle/Entity/Role.php:103]: Also think its important to have the key nullable know as I think there are synchronisations with e.g Active Directories which will only update the role `NAME` and not the key. So I think its important to have the key nullable and fallback to the name in the getIdentifier method.

File: src/Sulu/Bundle/SecurityBundle/Entity/Role.php [hunk 8]
```diff
@@ -93,14 +98,9 @@ public function getId()
         return $this->id;
     }
 
-    public function getRole()
-    {
-        return 'ROLE_SULU_' . strtoupper($this->name);
-    }
-
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Entity/Role.php:100]:
```suggestion

    /**
     * @deprecated Deprecated since Sulu 2.1 will be removed in 3.0 use "getIdentifier" instead.
     */
    public function getRole()
    {
        @trigger_error(sprintf('The "%s" method is deprecated since Sulu 2.1, use "%s" instead.', __METHOD__, 'getIdentifier'), E_USER_DEPRECATED)
        return $this->getIdentifier();
    }
```

File: src/Sulu/Bundle/SecurityBundle/Entity/Role.php (hunk 5 above)
Comment (nnatter) [reply src/Sulu/Bundle/SecurityBundle/Entity/Role.php:137]: i think so, according to the docs a role only *must start with `ROLE_`*, otherwise its an arbitrary string:
https://symfony.com/doc/current/security.html#roles

File: src/Sulu/Bundle/SecurityBundle/Resources/config/doctrine/Role.orm.xml (hunk 3 above)
Comment (nnatter) [reply src/Sulu/Bundle/SecurityBundle/Resources/config/doctrine/Role.orm.xml:12]: it is nullable 🙂

File: UPGRADE.md (hunk 6 above)
Comment (nnatter) [reply UPGRADE.md:14]: it is 🙂

File: UPGRADE.md (hunk 7 above)
Comment (nnatter) [reply UPGRADE.md:15]: it is not needed, but by doing this i dont need to fallback to the `name` in the `getIdentifier` method. but i am okay with implementing a fallback too

Comment (alexander-schranz) [reply UPGRADE.md:15]: I would definitely remove this and do a fallback else it could end strange for this who implemented synchronisation scripts.

Comment (nnatter) [reply UPGRADE.md:15]: alright 👍

File: src/Sulu/Bundle/SecurityBundle/Controller/RoleController.php [hunk 9]
```diff
@@ -242,12 +244,14 @@ public function putAction(Request $request, $id)
         /** @var RoleInterface $role */
         $role = $this->roleRepository->findRoleById($id);
         $name = $request->get('name');
+        $key = $request->get('key');
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Controller/RoleController.php:247]: We should always specify where to get the value: `$request->request->get('key')`.

File: src/Sulu/Bundle/SecurityBundle/Tests/Functional/Controller/RoleControllerTest.php [hunk 10]
```diff
@@ -494,15 +503,16 @@ public function testPutNotExisting()
         $this->assertStringContainsString('11230', $response->message);
     }
 
-    public function testPutWithExistingName()
+    public function testPutWithExistingKey()
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Tests/Functional/Controller/RoleControllerTest.php:506]: can we add a test where `'key' => null`

File: src/Sulu/Bundle/SecurityBundle/Exception/RoleKeyAlreadyExistsException.php [hunk 11]
```diff
@@ -12,31 +12,31 @@
 namespace Sulu\Bundle\SecurityBundle\Exception;
 
 /**
- * Exception is thrown when a Role is created or updated with an already existing name.
+ * Exception is thrown when a Role is created or updated with an already existing key.
  */
-class RoleNameAlreadyExistsException extends \Exception
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Exception/RoleKeyAlreadyExistsException.php:17]: we should just deprecate this one

File: src/Sulu/Bundle/SecurityBundle/Resources/config/doctrine/Role.orm.xml [hunk 12]
```diff
@@ -8,7 +8,8 @@
             <generator strategy="AUTO"/>
         </id>
 
-        <field name="name" type="string" column="name" length="60" unique="true"/>
+        <field name="name" type="string" column="name" length="60"/>
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/SecurityBundle/Resources/config/doctrine/Role.orm.xml:11]: Would still keep the uniquiness

File: UPGRADE.md [hunk 13]
```diff
@@ -2,6 +2,18 @@
...
 ## dev-master
 
+### Add key property to Role entity
+
+To allow for referencing `Role` entities by a human readable string, a `key` property was added to the `Role` entity.
+Furthermore, if the `key` is set, it is used for generating the identifier of a role instead of the `name` property.
+The following statements update the database to include the new property:
+
+```sql
+DROP INDEX UNIQ_13B749A05E237E06 ON se_roles;
```
Comment (alexander-schranz) [inline UPGRADE.md:12]: we should keep the uniqueness of the name

File: src/Sulu/Bundle/SecurityBundle/Exception/RoleKeyAlreadyExistsException.php (hunk 11 above)
Comment (alexander-schranz) [reply src/Sulu/Bundle/SecurityBundle/Exception/RoleKeyAlreadyExistsException.php:17]: if we keep uniquiness we should just keep or introduce a `RoleAlreadyExistsException`.

File: UPGRADE.md (hunk 13 above)
Comment (nnatter) [reply UPGRADE.md:12]: not sure about that 🤔 why would you keep it?

Comment (alexander-schranz) [reply UPGRADE.md:12]: Role Name think also from UX be unique.

Comment (alexander-schranz) [reply UPGRADE.md:12]: Also we still need to use the name and so should not change the behaviour is somebody rely on the RoleNameAlreadyExistsExcpetion

Comment (nnatter) [reply UPGRADE.md:12]: well i will trust you here 🙂

Comment (alexander-schranz) [reply UPGRADE.md]: I hope I don't disappoint you.

Comment (nnatter) [reply UPGRADE.md]: me too 😉
